File: src/utils/epg.ts
import { LiveChannel, ParsedProgram } from '../types/iptv';
import { findCurrentProgramIndex } from '../services/xmltv';

/**
 * Generates the EPG lookup key for a given channel.
 * For M3U, it strictly uses the epg_channel_id.
 * For Xtream, it falls back from epg_channel_id to stream_id.
 */
export const getEpgKey = (channel: LiveChannel, configType?: 'xtream' | 'm3u'): string => {
  if (configType === 'm3u') {
    return channel.epg_channel_id || '';
  }
  return channel.epg_channel_id || channel.stream_id?.toString() || '';
};

/**
 * Finds the currently airing program from an EPG array.
 */
export const getCurrentProgram = (
  epg: ParsedProgram[] | undefined,
  nowMs: number = Date.now()
): ParsedProgram | null => {
  if (!epg || epg.length === 0) return null;
  // ⚡ Bolt Optimization: Replace O(N) Array.findIndex with O(log N) binary search
  // Performance improvement: Drastically reduces lookup time in hot render loops,
  // preventing frame drops when displaying currently airing programs across large channel lists.
  const idx = findCurrentProgramIndex(epg, nowMs);
  return idx !== -1 ? epg[idx] : null;
};

File: flutter_app/lib/utils/epg.dart
import '../models/iptv.dart';

String getEpgKey(LiveChannel channel) {
  if (channel.epg_channel_id != null && channel.epg_channel_id!.isNotEmpty) {
    return channel.epg_channel_id!;
  }
  return channel.stream_id?.toString() ?? channel.name;
}

int findCurrentProgramIndex(List<ParsedProgram> programs, int nowTime) {
  if (programs.isEmpty) return -1;
  int low = 0;
  int high = programs.length - 1;

  while (low <= high) {
    int mid = low + ((high - low) >> 1);
    final prog = programs[mid];

    if (nowTime >= prog.start && nowTime < prog.end) {
      return mid;
    } else if (nowTime < prog.start) {
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }

  // fallback to full scan in case they aren't perfectly sorted
  for (int i = 0; i < programs.length; i++) {
    if (nowTime >= programs[i].start && nowTime < programs[i].end) {
      return i;
    }
  }
  return -1;
}

ParsedProgram? getCurrentProgram(List<ParsedProgram> programs, int nowTime) {
  int idx = findCurrentProgramIndex(programs, nowTime);
  return idx != -1 ? programs[idx] : null;
}
